library/src/WmiQuery.cpp

```diff
@@ -47,7 +47,7 @@ namespace
 		// <https://docs.microsoft.com/en-us/windows-hardware/drivers/install/identifiers-for-pci-devices>
 		// and insert wildcards for the subsystem vendor ID (if absent) and the device instance
 		return fmt::format(
-			L"PCI\\\\VEN_{:0>4X}&DEV_{:0>4X}&SUBSYS_{:0>4X}{}&REV_{:0>2X}%",
```

Comment:
It's been a while since I originally wrote this code (the git history of the private repository that was used for development prior to the public release dates this particular function to the 20th of June 2022), and although I can still remember the sequence of events that led to the addition of the conditional wildcard logic, I can no longer remember the specific rationale behind its implementation. From what I can recall, I tested the device query code by running it on a variety of on-premises machines and cloud VMs with GPUs from different vendors. On most systems, it functioned as expected, but I encountered one test machine on which the query failed to find any matching PnP devices. The conditional wildcard logic was introduced to resolve that failure.

Revisiting the code today and running some fresh tests, I'm left wondering what misguided logic lead me to implement the wildcard code in this manner. Here's the current state of things:

- As you've identified in your bugfix, the format string uses an incorrect width specifier for the subsystem ID value, which produces malformed strings and results in device matching failures. The format string specifies the minimum width but not the maximum, so subsystem ID values whose hexadecimal string representation does not contain any leading zeroes will produce a correct 8-digit output. Any values whose string representation would contain one or more leading zeroes will produce a malformed output of at least 4 digits and at most 7 digits.

- My comments indicate that the code checks for the presence of a subsystem vendor ID value and injects a wildcard into the output if the vendor ID is absent. However, the subsystem vendor ID is represented by the least-significant bytes of the subsystem ID value, and [the bitmask on line 44](https://github.com/TensorWorks/DirectX-Device-Plugins/blob/7ee9ff34a8003bc5038c881178aa097837a37536/library/src/WmiQuery.cpp#L44) instead isolates the most-significant bytes. (I can only assume that the bitmask reflects my actual intent from the time the code was written and the comments simply mislabel the target subset of the value, since an absence of non-zero bits in the least-significant positions would not impact the number of leading zeroes or the length of the output.) If the most-significant bytes are all zero then a wildcard is appended to the string representation of the least-significant bytes.

- In addition to the wildcard being entirely unnecessary in the first place, its positioning is wrong with respect to the condition that triggers it. In a scenario where the most-significant bytes are all zero and the respective leading zeroes in the hexadecimal string representation are omitted due to the incorrect width specifier, the missing digits precede those digits that are present. For the wildcard to function correctly, it would need to be prepended to the string representation of the least-significant bytes, rather than appended. The only scenario in which this would inadvertently function correctly is if both halves of the value are identical (i.e. the entire subsystem ID value is zero), and so I have to assume that this was the case on the test machine that motivated me to add the conditional wildcard logic in the first place.

Suffice to say, I will be removing the conditional wildcard code entirely now that the correct fix has been identified for the underlying logic bug.